src/domain/vocabulary/helper/dependencies.ts

```diff
@@ -0,0 +1,108 @@
+import * as E from 'fp-ts/Either'
+import * as S from 'fp-ts/string'
+import * as P from 'fp-ts/Predicate'
+import * as N from 'fp-ts/number'
+import * as ORD from 'fp-ts/Ord'
+import * as R from 'fp-ts/Record'
+import { pipe } from 'fp-ts/function'
+import type { Show } from 'fp-ts/lib/Show'
+import type { Deps } from '../command'
+import type { VocabularyPort } from '../port'
+
+export type Option = (deps: Deps) => E.Either<VocabularyDependenciesError, Deps>
+
+export const LanguageEmptyError = () =>
+  ({
+    _tag: 'language-empty'
+  } as const)
+
+/**
+ *  Error when withLanguage() function is called with an empty language argument
+ */
+export type LanguageEmptyError = ReturnType<typeof LanguageEmptyError>
+
+export const NegativeLimitError = (limit: number) =>
+  ({
+    _tag: 'negative-limit',
+    limit
+  } as const)
+
+/**
+ *  Error when withLimit() function is called with a negative limit argument
+ */
+export type NegativeLimitError = ReturnType<typeof NegativeLimitError>
+
+export const GatewayTypeError = (gateway: unknown) =>
+  ({
+    _tag: 'gateway-type',
+    gateway
+  } as const)
+
+/**
+ *  Error when withVocabularyGateway() function is called with a gateway that not satisfy the VocabularyPort constraint
+ */
+export type GatewayTypeError = ReturnType<typeof GatewayTypeError>
+
+export type VocabularyDependenciesError = LanguageEmptyError | NegativeLimitError | GatewayTypeError
+
+export const ShowVocabularyDependenciesError: Show<VocabularyDependenciesError> = {
+  show: (error: VocabularyDependenciesError): string => {
+    switch (error._tag) {
+      case 'language-empty': {
+        return `Error ${error._tag}: The given language parameter cannot be empty`
+      }
+      case 'negative-limit': {
+        return `Error ${error._tag}: The given limit parameter <${error.limit}> cannot be negative`
+      }
+      case 'gateway-type': {
+        return `Error ${error._tag}: The given gateway parameter <${JSON.stringify(
+          error.gateway
+        )}> does not implement at least the retrieveVocabulary function`
+      }
+    }
+  }
+}
+
+export const withVocabularyGateway =
+  (gateway: VocabularyPort): Option =>
+  (deps: Deps) =>
+    pipe(
+      gateway,
+      E.fromPredicate((g: VocabularyPort) => R.has('retrieveVocabulary', g), GatewayTypeError),
```

Comment:
Just because an object has a property doesn't mean that the property is of the correct type (I mean signature is correct). Reflexivity (or runtime type checking) is a bad practice. The type system of Typescript is designed to catch these kinds of errors at compile time, so you don't have to perform these checks at runtime.

Reply:
Right, so, if I understand well, there is no invariant to check for this function

Reply:
Yes. I don't see any preconditions to satisfy, other than those guaranteed by the typing system.